File: api-smartcard/ma3api-smartcard/src/tr/gov/tubitak/uekae/esya/api/smartcard/pkcs11/card/template/AladdinTemplate.java
package tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.card.template;

import java.util.ArrayList;
import java.util.List;

import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.CardType;
import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.card.ops.AladdinOps;
import tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.card.ops.IPKCS11Ops;

public class AladdinTemplate extends CardTemplate
{
	private static List<String> ATR_HASHES = new ArrayList<String>();
	
	static
	{
	    ATR_HASHES.add("3BD5180081313A7D8073C8211030");
	    ATR_HASHES.add("3BD518008131FE7D8073C82110F4");
	    ATR_HASHES.add("3B7F96000080318065B0846160FB120FFD829000");
	}
	
	public AladdinTemplate()
	{
		super(CardType.ALADDIN);
	}

	public synchronized IPKCS11Ops getPKCS11Ops()
	{
		if(mIslem == null)
			mIslem = new AladdinOps();
		return mIslem;
	}
	
	public String[] getATRHashes() 
	{
		return ATR_HASHES.toArray(new String[0]);
	}
	
	public static void addATR(String aATR)
	{
		ATR_HASHES.add(aATR);
	}
}

File: api-smartcard/ma3api-smartcard/csharp/src/pkcs11/card/template/AladdinTemplate.cs
using System;
using System.Collections.Generic;
using tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.card.ops;


namespace tr.gov.tubitak.uekae.esya.api.smartcard.pkcs11.card.template
{
    class AladdinTemplate : CardTemplate
    {
        protected static readonly List<String> ATR_HASHES = new List<String>();

        static AladdinTemplate()
	    {
	        ATR_HASHES.Add("3BD5180081313A7D8073C8211030");
            ATR_HASHES.Add("3BD518008131FE7D8073C82110F4");
            ATR_HASHES.Add("3B7F96000080318065B0846160FB120FFD829000");
        }

        public AladdinTemplate(CardType aCardType)
            : base(aCardType)
        {

        }

        public override IPKCS11Ops getPKCS11Ops() 
	    {
            lock (PKCS11GetLock)
            {
                if (mIslem == null)
                    mIslem = new AladdinOps();
                return mIslem;
            }
        }

        public override String[] getATRHashes()
        {
            return ATR_HASHES.ToArray();
        }

        public void addATR(String aATR)
        {
            ATR_HASHES.Add(aATR);
        }
    }
}
